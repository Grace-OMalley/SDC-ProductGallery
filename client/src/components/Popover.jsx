
import React from 'react';
import ImageRow from './ImageRow.jsx';

class Popover extends React.Component {

  constructor (props) {
    super(props);
    this.state = {
      selected: props.selected || 0,
    };
    this.imageContainer = React.createRef();
    this.popover = React.createRef();
    this.image = React.createRef();
  }

  mouseHandler(index) {
    this.setState({selected: +index.currentTarget.dataset.index});
  }

  moveHandler(e) {
    const imageContainer = this.imageContainer.current;
    const image = this.image.current;
    const popover = this.popover.current;
    let containerHeight = imageContainer.offsetHeight;
    let containerWidth = imageContainer.offsetWidth;
    let yZoom = image.naturalHeight > containerHeight;
    let xZoom = image.naturalWidth > containerWidth;
    imageContainer.classList.toggle('zoomable', xZoom || yZoom);
    if (!imageContainer.classList.contains('zoomed')) {
      return;
    }
    let marginTop = 0;
    let marginLeft = 0;
    if (yZoom) {
      let y = e.clientY - (popover.offsetTop + imageContainer.offsetTop);
      let percentage = image.offsetHeight - containerHeight;
      let range = (( y / containerHeight ) - 0.25) * 2;
      range = range < 0 ? 0 : range > 1 ? 1 : range;
      marginTop = percentage - (2 * range * percentage) + 'px';
    }
    if (xZoom) {
      let x = e.clientX - (popover.offsetLeft + imageContainer.offsetLeft);
      let percentage = image.offsetWidth - containerWidth;
      let range = (( x / containerWidth ) - 0.25) * 2;
      range = range < 0 ? 0 : range > 1 ? 1 : range;
      marginLeft = percentage - (2 * range * percentage) + 'px';
    }
    image.style.margin = marginTop + ' 0 ' + '0 ' + marginLeft;
  }

  render() {
    const imageRows = [];

    for (let i = 0; i < this.props.images.length; i++) {
      const index = Math.floor(i / 4);
      if (!Array.isArray(imageRows[index])) {
        imageRows[index] = [];
      }
      imageRows[index].push(this.props.images[i]);
    }

    return (
      this.props.visible && <div className="overlay">
        <div ref={this.popover} className="popoverGallery" onMouseMove={(e) => this.moveHandler(e)}>
          <button onClick={this.props.closeHandler} className="popoverClose">Ã—</button>
          <div ref={this.imageContainer} className="largeImage" onClick={(e) => {
            let imageContainer = this.imageContainer.current;
            if (imageContainer.classList.contains('zoomable')) {
              imageContainer.classList.toggle('zoomed');
              if (!imageContainer.classList.contains('zoomed')) {
                this.image.current.style.margin = 0;
              }
              this.moveHandler(e);
            }
          }}>
            <img ref={this.image} src={this.props.images[this.state.selected]} />
          </div>
          <div className="imageColumn">
            <div className="imageColumnTitle">{this.props.name}</div>
            {imageRows.map((images, offset) =>
              <ImageRow images={images} key={offset} offset={offset * 4} mouseHandler={() => {}} clickHandler={(e) => this.mouseHandler(e)} selected={this.state.selected}/>
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default Popover;