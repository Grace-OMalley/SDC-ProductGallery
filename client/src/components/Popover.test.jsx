import React from 'react';
import renderer from 'react-test-renderer';
import {cleanup, waitFor, waitForElementToBeRemoved, fireEvent, render, queryHelpers, buildQueries, screen} from '@testing-library/react';
import Popover from './Popover';

afterEach(cleanup);

it('renders correctly', () => {
  const tree = renderer
    .create(<Popover
      images={[1, 2, 3]}
      visible={true}
      name={'test'}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('is invisible when passed visible=false', () => {
  const {container} = render(
    <Popover
      images={[1, 2, 3]}
      visible={false}
      name={'test'}
    />
  );
  const overlay = container.querySelector('.overlay');
  expect(overlay).toBeNull();
});

it('changes image when imageList is clicked', () => {
  const {container} = render(
    <Popover
      images={[1, 2, 3]}
      visible={true}
      name={'test'}
    />
  );
  const image = container.querySelector('.largeImage img');
  expect(image.getAttribute('src')).toBe('1');
  const image1 = container.querySelector('[data-index="0"]');
  const image2 = container.querySelector('[data-index="1"]');
  fireEvent.click(image2);
  expect(image.getAttribute('src')).toBe('2');
});

it('passes close function to closebutton', async () => {
  const closeHandler = jest.fn();
  const {container} = render(
    <Popover
      images={[1, 2, 3]}
      closeHandler={closeHandler}
      visible={true}
      name={'test'}
    />
  );
  const close = container.querySelector('.popoverClose');
  fireEvent.click(close);
  expect(closeHandler).toHaveBeenCalledTimes(1);
});

it('doesn\'t crash when user moves mouse over main-image', () => {
  const {container} = render(
    <Popover
      images={[1, 2, 3]}
      visible={true}
      name={'test'}
    />
  );
  const image = container.querySelector('.largeImage img');
  fireEvent.mouseMove(image);
});

it('toggles image zoom / pan', () => {
  const {container, rerender} = render(
    <Popover
      images={[1, 2, 3]}
      visible={true}
      name={'test'}
    />
  );
  let image = container.querySelector('.largeImage');
  //test clicking when image is 'non-zoomable'
  expect(image.classList).not.toContain('zoomable');
  fireEvent.click(image);
  //setup for test
  rerender(<Popover
    images={[1, 2, 3]}
    visible={true}
    name={'test'}
    test={true}
  />);
  fireEvent.mouseMove(image);
  expect(image.classList).toContain('zoomable');
  expect(image.classList).not.toContain('zoomed');
  fireEvent.click(image);
  expect(image.classList).toContain('zoomed');
  fireEvent.click(image);
  expect(image.classList).not.toContain('zoomed');
});