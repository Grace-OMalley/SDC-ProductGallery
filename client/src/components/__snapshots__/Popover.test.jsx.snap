// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="overlay"
>
  <div
    className="popoverGallery"
  >
    <button
      className="popoverClose"
    >
      Ã—
    </button>
    <div
      className="largeImage"
      onClick={[Function]}
      onMouseMove={[Function]}
    >
      <img
        src={1}
      />
    </div>
    <div
      className="imageColumn"
    >
      <div
        className="imageColumnTitle"
      >
        test
      </div>
      <ul
        className="imageRow"
      >
        <li
          className="selected"
          data-index={0}
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <img
            src={1}
          />
        </li>
        <li
          className=""
          data-index={1}
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <img
            src={2}
          />
        </li>
        <li
          className=""
          data-index={2}
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <img
            src={3}
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;
